engine/runtime/src/test/java/org/enso/interpreter/test/AtomConstructorTest.java

```diff
@@ -0,0 +1,198 @@
+package org.enso.interpreter.test;
+
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertTrue;
+import static org.junit.Assert.fail;
+
+import java.util.function.Function;
+import org.enso.interpreter.runtime.data.atom.Atom;
+import org.enso.interpreter.runtime.data.atom.AtomConstructor;
+import org.enso.interpreter.runtime.data.atom.AtomNewInstanceNode;
+import org.enso.interpreter.runtime.data.atom.StructsLibrary;
+import org.enso.interpreter.runtime.error.PanicException;
+import org.enso.polyglot.MethodNames;
+import org.graalvm.polyglot.Context;
+import org.junit.AfterClass;
+import org.junit.BeforeClass;
+import org.junit.Test;
+
+/**
+ * @author devel
+ */
+public class AtomConstructorTest extends TestBase {
+
+  private static Context ctx;
+
+  public AtomConstructorTest() {}
+
+  @BeforeClass
+  public static void initContext() {
+    ctx = createDefaultContext();
+  }
+
+  @AfterClass
+  public static void closeContext() {
+    ctx.close();
+  }
+
+  @Test
+  public void newInstance() {
+    var code = """
+        type NoPrime
+            A a b c
+        """;
+    var module = ctx.eval("enso", code);
+    var consA = module.invokeMember(MethodNames.Module.EVAL_EXPRESSION, "NoPrime.A");
+    var raw = unwrapValue(ctx, consA);
+
+    assertTrue("It is atom constructor: " + raw, raw instanceof AtomConstructor);
+    var cons = (AtomConstructor) raw;
+
+    Function<Object[], Atom> factory =
+        args -> AtomNewInstanceNode.getUncached().newInstance(cons, args);
+    assertAtomFactory("AtomConstructor.newInstance without priming", factory);
+    assertLessArguments("AtomConstructor.newInstance without priming", factory);
+  }
+
+  @Test
+  public void newAtomNewInstanceNode() {
+    var code = """
+        type X
+            A a b c
+        """;
+    var module = ctx.eval("enso", code);
+    var xA = module.invokeMember(MethodNames.Module.EVAL_EXPRESSION, "X.A");
+    var raw = unwrapValue(ctx, xA);
+
+    assertTrue("It is atom constructor: " + raw, raw instanceof AtomConstructor);
+    var cons = (AtomConstructor) raw;
+
+    var node = AtomNewInstanceNode.create();
+    Function<Object[], Atom> factory = args -> node.newInstance(cons, args);
+    assertAtomFactory("AtomNewInstanceNode.create", factory);
+    assertLessArguments("AtomNewInstanceNode.create", factory);
+
+    assertAtomFactory("AtomConstructor.newInstance with priming", factory);
```

Comment:
There are two tests. First one verifies the behavior of `getUncached()` in isolation. The second one checks whether both _cached_ and _uncached_ variants co-operate well.